heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java

```diff
@@ -39,12 +40,105 @@
 import com.twitter.heron.spi.utils.ShellUtils;
 
 public final class SchedulerUtils {
-  public static final int PORTS_REQUIRED_FOR_EXECUTOR = 9;
+  public static final int PORTS_REQUIRED_FOR_EXECUTOR = ExecutorPorts.getRequiredNumOfPorts();
   public static final int PORTS_REQUIRED_FOR_SCHEDULER = 1;
   public static final String SCHEDULER_COMMAND_LINE_PROPERTIES_OVERRIDE_OPTION = "P";
 
   private static final Logger LOG = Logger.getLogger(SchedulerUtils.class.getName());
 
+  /**
+   * Enum that defines the type of ports that an heron executor needs
+   */
+  public static class ExecutorPorts extends HashMap<ExecutorPorts.ExecutorPortNames, Object> {
+    private static final long serialVersionUID = 3594142561871208879L;
+
+    public enum ExecutorPortNames {
+      MASTER_PORT("master"),
+      TMASTER_CONTROLLER_PORT("tmaster-ctl"),
+      TMASTER_STATS_PORT("tmaster-stats"),
+      SHELL_PORT("shell-port"),
+      METRICS_MANAGER_PORT("metrics-mgr"),
+      SCHEDULER_PORT("scheduler"),
+      METRICS_CACHE_MASTER_PORT("metrics-cache-m"),
+      METRICS_CACHE_STATS_PORT("metrics-cache-s"),
+      CHECKPOINT_MANAGER_PORT("ckptmgr");
+
+      private final String name;
+
+      ExecutorPortNames(String name) {
+        this.name = name;
+      }
+
+      public String toString() {
+        return this.name;
+      }
+
+      public static String[] getPortNames() {
+        List<String> names = new LinkedList<>();
+        for (ExecutorPortNames executorPortNames : ExecutorPortNames.values()) {
+          names.add(executorPortNames.toString());
+        }
+
+        return (String[]) names.toArray();
+      }
+    }
+
+    public static String[] getRequiredPorts() {
```

Comment:
Is this method duplicating the responsibility of `getPortNames()`

Reply:
Currently it contains all the defined port types, but this is not going to be the case in the future.  For example, the ports for remote debugging as part of https://github.com/twitter/heron/pull/2484 is optional and with not be part of the required.  And currently we pass in all the ports that any possible executor will need which is also unnecessary.  For instance only container-0 need the ports related to the tmaster.  That cleanup can be done in a follow up PR.

Reply:
I am looking at #2484. It seems remote debugging ports are required by the scripts and proto files. If that is correct, are you planning to change that? 

It also seems remote debugging ports will get promoted to `required` if remote debugging is enabled?

Reply:
remote debugging ports will only be need if the user sets a flag in there topology.  This will be an optional port .  The scheduler will decide how to get these ports and pass them to SchedulerUtils.